### PR TITLE
(PDB-1252) Remove clamq; use ActiveMQ directly

### PR DESCRIPTION
**Pull request by rbrw**
(PDB-1252) Remove clamq; use ActiveMQ directly
    
Remove clamq entirely, and instead, just call ActiveMQ and javax.jms
directly.

### CODE REVIEW

**Comment by grimradical** [comment]
This is good stuff, and long overdue.

Minor point: I think in a few spots you refer to JMX (management/metrics extensions) when I think you mean JMS (messaging). :)

**Comment by rbrw** [comment]
@grimradical Oh, right -- fixed.  And thanks.

**Comment by rbrw** [comment]
@pljenkinsro retest this please

**Comment by wkalt** [comment]
@pljenkinsro retest this please

File: project.clj [hunk 1]
```diff
@@ -38,7 +38,7 @@
                  [clj-stacktrace "0.2.6"]
                  [metrics-clojure "0.7.0" :exclusions [org.clojure/clojure org.slf4j/slf4j-api]]
                  [clj-time "0.5.1"]
-                 [org.clojure/java.jmx "0.2.0"]
+                 [org.clojure/java.jmx "0.3.0"]
```
Comment (mullr) [inline project.clj]: 0.3.1 was just released a few days ago, we should probably use that.

File: project.clj [hunk 2]
```diff
@@ -55,10 +55,14 @@
...
                  ;; MQ connectivity
-                 [clamq/clamq-activemq "0.4" :exclusions [org.slf4j/slf4j-api]]
-                 [org.apache.activemq/activemq-core "5.7.0" :exclusions [org.slf4j/slf4j-api org.fusesource.fuse-extra/fusemq-leveldb]]
+                 [org.apache.activemq/activemq-core "5.7.0"
+                  :exclusions [org.slf4j/slf4j-api
+                               org.fusesource.fuse-extra/fusemq-leveldb]]
+                 [org.apache.activemq/activemq-pool "5.7.0"
+                  :exclusions [org.slf4j/slf4j-api]]
+
                  ;; bridge to allow some spring/activemq stuff to log over slf4j
```
Comment (mullr) [inline project.clj]: Is the 'spring' part of this still accurate?

File: src/puppetlabs/puppetdb/command.clj [hunk 3]
```diff
@@ -131,27 +132,29 @@
           (string? mq-endpoint)
           (string? raw-command)]
    :post [(string? %)]}
+  ;; FIXME: is this efficient enough?  It's rebuilding the whole MQ
```
Comment (mullr) [inline src/puppetlabs/puppetdb/command.clj]: This is what we talked about before, right? I'd love to see this fixed, but I'd be totally fine with making it a separate ticket.o

File: project.clj (hunk 2 above)
Comment (rbrw) [inline project.clj]: @mullr They were when I made these changes, but I'll be checking again before the final version.

File: src/puppetlabs/puppetdb/command.clj [hunk 4]
```diff
@@ -131,27 +132,29 @@
...
-      (mq/connect-and-publish! conn mq-endpoint raw-command {"received" (kitchensink/timestamp)
-                                                             "id" id}))
+    (with-open [factory (mq/activemq-connection-factory mq-spec)]
+      (mq/connect-and-publish! factory mq-endpoint raw-command
+                               {"received" (kitchensink/timestamp)
+                                "id" id}))
     id))
 
-(defn enqueue-command!
+(defn-validated enqueue-command! :- s/Str
```
Comment (mullr) [inline src/puppetlabs/puppetdb/command.clj]: Question: when do we use defn-validated in this codebase? What's the motivation for using it here? Don't read too much in to this, I just want to know.

File: src/puppetlabs/puppetdb/command.clj (hunk 3 above)
Comment (rbrw) [inline src/puppetlabs/puppetdb/command.clj]: @mullr Already fixed here -- I've reworked all of that side to share a single connection.

Comment (wkalt) [inline src/puppetlabs/puppetdb/command.clj]: comment needs to be removed at any rate

File: src/puppetlabs/puppetdb/cli/services.clj [hunk 5]
```diff
@@ -204,30 +202,43 @@
...
+          (log/error "waited for updater thread for 3s but it didn't finish"))
+        (catch java.util.concurrent.CancellationException ex
+          true)))))
+
 (defn stop-puppetdb
-  "Callback to execute any necessary cleanup during a normal shutdown."
   [context]
+  ;; FIXME: wait for in-flight jetty calls, and then block access
+  ;; here, say via wrap-with-puppetdb-middleware returning 503 or
+  ;; similar? (preexisting condition)
```
Comment (rbrw) [inline src/puppetlabs/puppetdb/cli/services.clj]: I've filed a separate ticket for this (PDB-1388).

**Comment by rbrw** [comment]
@pljenkinsro retest this please

**Comment by rbrw** [comment]
@mullr @senior hopefully I've addressed your comments, and I've squashed all the commits.  The only remaining issues are the three "Q:" comments regarding exception handling.

**Comment by rbrw** [comment]
@pljenkinsro retest this please

**Comment by rbrw** [comment]
Note that there was some additional discussion here:  https://github.com/rbrw/puppetdb/commit/ec57e88b2aea86e09b226d976a733b1a515c8d9e#commitcomment-10566369

File: src/puppetlabs/puppetdb/cli/services.clj [hunk 6]
```diff
@@ -78,15 +79,10 @@
 (def mq-addr "vm://localhost?jms.prefetchPolicy.all=1&create=false")
 (def mq-endpoint "puppetlabs.puppetdb.commands")
 
-(defn send-command!
-  "Send the command using the default MQ address and endpoint"
-  [command version payload]
-  (command/enqueue-command! mq-addr mq-endpoint command version payload))
-
 (defn auto-deactivate-nodes!
```
Comment (wkalt) [inline src/puppetlabs/puppetdb/cli/services.clj]: this is being called with a partial below as if it still only had one argument. I think that's what's causing the test failues.

Comment (wkalt) [inline src/puppetlabs/puppetdb/cli/services.clj]: looks like you can just change it to #(auto-deactivate-nodes! node-ttl % mq-connection) and everything will work.

File: src/puppetlabs/puppetdb/command.clj [hunk 7]
```diff
@@ -119,40 +120,30 @@
...
-  annotating."
-  [mq-spec mq-endpoint raw-command]
-  {:pre  [(string? mq-spec)
-          (string? mq-endpoint)
-          (string? raw-command)]
-   :post [(string? %)]}
+(defn-validated enqueue-raw-command! :- s/Str
+  "Submits raw-command to the mq-endpoint of mq-connection and returns
+  its id. Annotates the command via annotate-command, unless it can't
+  be parsed."
```
Comment (wkalt) [inline src/puppetlabs/puppetdb/command.clj]: does this actually do anything with annotation?

File: src/puppetlabs/puppetdb/mq.clj [hunk 8]
```diff
@@ -142,81 +136,107 @@
...
-    (mq-consumer/start consumer)
-    (while (> limit (count @contents))
-      (Thread/sleep 10))
-
-    (mq-consumer/close consumer)
-    (if (realized? mq-error)
-      (throw @mq-error)
-      (vec (take limit @contents)))))
+;; In the final patch, perhaps move the code here, but this
+;; should make review easier.
```
Comment (wkalt) [inline src/puppetlabs/puppetdb/mq.clj]: marking for visibility

File: src/puppetlabs/puppetdb/cli/services.clj (hunk 6 above)
Comment (rbrw) [inline src/puppetlabs/puppetdb/cli/services.clj]: @wkalt of course, thanks for tracking that down.

File: src/puppetlabs/puppetdb/command.clj (hunk 7 above)
Comment (rbrw) [inline src/puppetlabs/puppetdb/command.clj]: @wkalt ahh, good catch.

File: src/puppetlabs/puppetdb/cli/services.clj [hunk 9]
```diff
@@ -329,8 +332,10 @@
                                   ;; anything referencing an env or resource
                                   ;; param is purged first
                                   garbage-collect!]]
-
-        (gc-task #(apply perform-db-maintenance! write-db (remove nil? db-maintenance-tasks)))
+        ;; Run database maintenance tasks seqentially to avoid
+        ;; competition. Each task must handle its own errors.
+        (gc-task #(doseq [maintain (remove nil? db-maintenance-tasks)]
+                    (maintain write-db)))
```
Comment (rbrw) [inline src/puppetlabs/puppetdb/cli/services.clj]: @wkalt does this look OK to you, or think I should revert it?  (I thought it was a bit more straightforward.)

**Comment by rbrw** [comment]
@pljenkinsro retest this please

File: src/puppetlabs/puppetdb/cli/services.clj (hunk 9 above)
Comment (wkalt) [inline src/puppetlabs/puppetdb/cli/services.clj]: @rbrw I agree, this seems nicer to me.

**Comment by wkalt** [comment]
This causes an ugly JMS error to get logged (to the log, not the console) every time puppetdb is stopped. You can reproduce by tailing the log and killing PDB with kill or ctrl-c.

File: src/puppetlabs/puppetdb/mq_listener.clj [hunk 10]
```diff
@@ -393,31 +331,72 @@
...
+        handle (create-message-handler send discard-dir process-msg)]
+    (.setMessageListener
+     consumer
+     (reify MessageListener
+       (onMessage [this msg]
+         (try
+           (mq/commit-or-rollback sess (handle (mq/convert-jms-message msg)))
+           (catch Throwable ex
+             ;; Q: What do we really want to do here?
+             (log/error ex "message receive failed")
```
Comment (wkalt) [inline src/puppetlabs/puppetdb/mq_listener.clj]: I think all the Q: comments can be removed without code changes, unless there's something you think should be done in these cases aside from logging the exception.

Comment (rbrw) [inline src/puppetlabs/puppetdb/mq_listener.clj]: @wkalt done, thanks.

**Comment by wkalt** [comment]
I'm +1 after tests pass; looks like @mullr's comment about spring also needs resolution.

**Comment by wkalt** [comment]
@pljenkinsro retest this please

File: project.clj (hunk 2 above)
Comment (rbrw) [inline project.clj]: @mullr ahh, so I thought you were talking about the exclusions, which I did double-check that we need.  With respect to the comment, I suppose it's accurate in the sense that we do still pull in some spring bits via activemq.

**Comment by mullr** [comment]
:+1: from me

**Pull request merged**